drivers/nrf24l01p_ng/nrf24l01p_ng_communication.c

```diff
@@ -0,0 +1,220 @@
+ *
+ * Suppose you want to give pipe 0 the address 01:02:03:04.
+ * If you send:
+ * 01 ---> [ x,  x,  x,  x, 01]
+ * 02 ---> [ x,  x,  x, 02, 01]
+ * 03 ---> [ x,  x, 03, 02, 02]
+ * 04 ---> [ x, 04, 03, 02, 01]
+ *
+ * The transceiver does not have the address you expected.
+ * But over SPI you would read:
+ * [ x, 04, 03, 02, 01] --->  01:02:03:04
+ * because it sends the LSByte first. But it will not
+ * receice frames that are actually addressed to 01:02:03:04.
+ *
+ * So if you want to assign the address 01:02:03:04
+ * you first have to swap the bytes, before you send data over SPI.
+ *
+ */
+static
+void _swap_bytes(uint8_t* buffer, size_t len) {
+    for (size_t i = 0; i < len / 2; i++) {
+        uint8_t tmp = buffer[i];
+        buffer[i] = buffer[len - 1 - i];
+        buffer[len - 1 - i] = tmp;
+    }
+}
+
+int nrf24l01p_ng_acquire(nrf24l01p_ng_t *dev)
+{
+    return spi_acquire(dev->params.spi, dev->params.pin_cs, SPI_MODE_0,
+                       dev->params.spi_clk);
+}
+
+void nrf24l01p_ng_release(nrf24l01p_ng_t *dev)
+{
+    spi_release(dev->params.spi);
+}
+
+uint8_t nrf24l01p_ng_read_reg(const nrf24l01p_ng_t *dev, uint8_t reg,
+                              uint8_t *dest, size_t len)
+{
+    uint8_t cmd = NRF24L01P_NG_CMD_R_REGISTER(reg);
+    uint8_t status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+    spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, NULL, dest, len);
+    if (len > 1) {
+        _swap_bytes(dest, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_ng_write_reg(const nrf24l01p_ng_t *dev, uint8_t reg,
+                               const uint8_t *src, size_t len)
+{
+    uint8_t cmd = NRF24L01P_NG_CMD_W_REGISTER(reg);
+    uint8_t status;
+    if (len > 1) {
+        len = MIN(len, NRF24L01P_NG_REG_MAX_WIDTH);
+        uint8_t lsrc[NRF24L01P_NG_REG_MAX_WIDTH];
+        memcpy(lsrc, src, len);
+        _swap_bytes(lsrc, len);
+        status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+        spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, lsrc, NULL, len);
+    }
+    else {
+        status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+        spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, src, NULL, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_ng_read_rx_payload(const nrf24l01p_ng_t *dev, void *dest,
+                                     size_t len)
+{
+    uint8_t cmd = NRF24L01P_NG_CMD_R_RX_PAYLOAD;
+    uint8_t status = spi_transfer_byte(SPI_BUS, SPI_PIN_CS, true, cmd);
+    spi_transfer_bytes(SPI_BUS, SPI_PIN_CS, false, &cmd, dest, len);
+    if (len > 1) {
+        _swap_bytes(dest, len);
+    }
+    return status;
+}
+
+uint8_t nrf24l01p_ng_write_tx_payload(const nrf24l01p_ng_t *dev,
+                                      const void *src, size_t len)
+{
+    uint8_t cmd = NRF24L01P_NG_CMD_W_TX_PAYLOAD;
+    uint8_t status;
+    if (len > 1) {
+        len = MIN(len, NRF24L01P_NG_MAX_PAYLOAD_WIDTH);
+        uint8_t lsrc[NRF24L01P_NG_MAX_PAYLOAD_WIDTH];
```

Comment:
Should we really truncate payloads silently?

Reply:
I should better make an assert that `len <= NRF24L01P_NG_MAX_PAYLOAD_WIDTH`.